tools/plagiarism-checker/src/index.ts

```diff
@@ -0,0 +1,93 @@
+import { Hono } from "hono"
+
+interface SearchResult {
+  title?: string | null
+  link?: string | null
+}
+
+interface SearchResponse {
+  items: Array<SearchResult>
+}
+
+interface PlagiarismResults {
+  sentence: string
+  matches: Array<SearchResult>
+}
+
+type Env = {
+  GOOGLE_API_KEY: string
+  GOOGLE_SEARCH_ENGINE_CX: string
+}
+
+const app = new Hono<{ Bindings: Env }>()
+
+app.use("*", async (c, next) => {
+  const apiKey = c.env.GOOGLE_API_KEY
+  if (!apiKey) {
+    return c.text("API key not found", 400)
+  }
+
+  const searchEngine = c.env.GOOGLE_SEARCH_ENGINE_CX
+  if (!searchEngine) {
+    return c.text("Search engine not specified", 400)
+  }
+
+  return next()
+})
+
+app.post("/", async (c) => {
+  const { text } = await c.req.json()
+
+  if (typeof text !== "string") {
+    return c.json({ error: "Invalid JSON format" }, 400)
+  }
+
+  const sentences = text
+    .split(".")
+    // if it's less than 10 characters - we don't treat it as a sentence and skip it
+    .filter((sentence) => sentence.trim().length > 10)
```

Comment:
1. What about egde cases where a sentence inside contains abbreviations, like `U.S`, `e.g.`, `etc.` ? Maybe consider an alternative approach to splitting.
2. How did you come up with a 10-character limit?

Reply:
> 1. What about egde cases where a sentence inside contains abbreviations, like `U.S`, `e.g.` ? Maybe consider an alternative approach to splitting.
> 
>     2. How did you come up with a 10-character limit?

Basically 10-character limit is meant to filter out abbreviations and irrelevant pieces of text. We assume that anything less then 10 characters can't be properly checked for a plagiarism with our method, because it's probably 1-2 words and google will surely get some matches, even if the text is actually unique. I can see how that's not ideal.

Another way would be to take in every N words instead of a sentence. But if 1 sentence is copy-pasted from the web and the next one is not - they are gonna get mixed up and plagiarism will not be found (since we are looking for 100% matches with google api).

We could use a library for splitting sentences, but I found this cool [regex](https://stackoverflow.com/a/72280712/10957343) that I am going to use instead.  

Now another problem I found in the process is that it treats headers as part of the following sentence: (so did my split(".") method)
![image](https://github.com/1712n/dn-institute/assets/45823609/d081579d-d9f0-4876-9ea7-cfcd04ee6b44) 
To fix that, I added another regex to filter out headers. 

And I removed the 10-char min limit.